spec/javascripts/components/table-filter-spec.js

```diff
@@ -0,0 +1,100 @@
+            '</tr>' +
+            '<tr class="govuk-table__row">' +
+              '<td>klm</td>' +
+              '<td>nop</td>' +
+              '<td>£35</td>' +
+            '</tr>' +
+          '</tbody>' +
+        '</table>' +
+        '<p class="govuk-table__message govuk-table__message--hidden">That search returns no results.</p>' +
+      '</div>'
+
+    window.setFixtures(FIXTURE)
+
+    element = document.querySelector('[data-module="table-filter"]')
+    searchInput = document.querySelector('input')
+    tableFilter = new GOVUK.Modules.TableFilter(element)
+  })
+
+  it('Calls the `updateRows` method when the user types in the input field', function () {
+    var event = new window.Event('input')
+
+    spyOn(tableFilter, 'updateRows')
+    tableFilter.init()
+    searchInput.dispatchEvent(event)
+
+    expect(tableFilter.updateRows).toHaveBeenCalledTimes(1)
+  })
+
+  it('Sets the correct classes when the `updateRows` method is called', function () {
+    var rows = document.querySelectorAll('.govuk-table__row')
+    var message = document.querySelector('.govuk-table__message')
+    var hiddenRowClass = tableFilter.hiddenRowClass
+    var hiddenMessageClass = tableFilter.hiddenMessageClass
+
+    // User enters 'b'
+    searchInput.value = 'b'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(false)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(true)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(true)
+
+    // User enters 'o'
+    searchInput.value = 'o'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(false)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(false)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(true)
+
+    // User enters 'a'
+    searchInput.value = 'a'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(true)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(true)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(false)
+
+    // User enters 'op'
+    searchInput.value = 'op'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(true)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(false)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(true)
+
+    // User enters 'OP'
+    searchInput.value = 'OP'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(true)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(false)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(true)
+
+    // User enters '£3'
+    searchInput.value = '£3'
+    tableFilter.updateRows()
+    expect(rows[0].classList.contains(hiddenRowClass)).toBe(true)
+    expect(rows[1].classList.contains(hiddenRowClass)).toBe(false)
+    expect(message.classList.contains(hiddenMessageClass)).toBe(true)
+  })
```

Comment:
Should there also be a test that checks that everything reappears if the user clears the content of the filter/search input?

Reply:
I’m not sure about this. I would say that scenario is already covered by the existing instances and I have also been removing unnecessary and over-zealous testing. Very happy to discuss the test coverage and how we should approach that more broadly in general though (see below).

Reply:
Thanks for adding that. Unfortunately I think I can foresee a problem. Correct me if I'm wrong, but it looks like the general state of the test fixture isn't reset after each test - you're using a `beforeAll` to setup things once. Since the tests run in a random order, it's entirely possible that the reset test you've added could be run first, in which case it doesn't do anything - the input value is already empty.

For this specific test I'd recommend doing a search first, checking the results are as expected, then clearing the input and checking that everything has been reset to normal.

As a general strategy I'd also suggest that your tests be as self contained as possible to avoid race conditions that could incorrectly pass or fail following tests. Usually we recreate the fixture and any associated variables for each test using a `beforeEach` to ensure that the state of the code is as known as possible for each test.